### PR TITLE
[Please Merge] Php like UI

### PR DESCRIPTION
**Pull request by KugliLuca**
Added a like button to the unguard UI as well as changes to the Like-API and Microblog-API.

### CODE REVIEW

File: src/frontend/.env [hunk 1]
```diff
@@ -7,7 +7,7 @@ PROXY_SERVICE_ADDRESS=localhost:8081
 AD_SERVICE_ADDRESS=localhost:8082
 STATUS_SERVICE_ADDRESS=localhost:8083
 USER_AUTH_SERVICE_ADDRESS=localhost:9091
-LIKE_SERVICE_ADDRESS=localhost:8000
+LIKE_SERVICE_ADDRESS=unguard-like-service
```
Comment (Dirli-V) [inline src/frontend/.env]: [NEED_INFO] Why was this changed?

File: src/frontend/app.js [hunk 2]
```diff
@@ -62,6 +63,7 @@ frontendLogger.info("FRONTEND_BASE_PATH is set to " + process.env.FRONTEND_BASE_
 frontendLogger.info("AD_SERVICE_BASE_PATH is set to " + process.env.AD_SERVICE_BASE_PATH);
 frontendLogger.info("STATUS_SERVICE_BASE_PATH is set to " + process.env.STATUS_SERVICE_BASE_PATH);
 frontendLogger.info("PROFILE_SERVICE_ADDRESS is set to " + process.env.PROFILE_SERVICE_ADDRESS);
+frontendLogger.info("LIKE_SERVICE_ADDRESS is set to " + process.env.UNGUARD_LIKE_SERVICE_SERVICE_HOST + ":" + process.env.UNGUARD_LIKE_SERVICE_SERVICE_PORT);
```
Comment (Dirli-V) [inline src/frontend/app.js]: [DISCUSS] I guess "LIKE_SERVICE_ADDRESS" be easier and more in line with the others?

File: src/frontend/app.js [hunk 3]
```diff
@@ -142,13 +144,16 @@ app.use((req, res, next) => {
     const MEMBERSHIP_SERVICE_API = createAxiosInstance(req, "http://" + process.env.MEMBERSHIP_SERVICE_ADDRESS + process.env.MEMBERSHIP_SERVICE_BASE_PATH, membershipServiceApiLogger, cookieHeader)
     const PROFILE_SERVICE_API = createAxiosInstance(req, "http://" + process.env.PROFILE_SERVICE_ADDRESS, profileServiceLogger);
     const STATUS_SERVICE_API =createAxiosInstance(req, "http://" + process.env.STATUS_SERVICE_ADDRESS + process.env.STATUS_SERVICE_BASE_PATH, statusServiceApiLogger);
+    const LIKE_SERVICE_API = createAxiosInstance(req, "http://" + process.env.UNGUARD_LIKE_SERVICE_SERVICE_HOST + ":" + process.env.UNGUARD_LIKE_SERVICE_SERVICE_PORT, likeServiceLogger, cookieHeader);
+
```
Comment (Dirli-V) [inline src/frontend/app.js:148]: [NIT] superfluous new line

File: src/frontend/site.js [hunk 4]
```diff
@@ -14,10 +14,10 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-const { handleError, statusCodeForError } = require("./controller/errorHandler");
-const { getJwtUser, hasJwtRole, getJwtUserId } = require('./controller/cookie');
-const { roles } = require('./model/role');
-const { extendURL, extendRenderData } = require("./controller/utilities.js");
+const {handleError, statusCodeForError} = require("./controller/errorHandler");
```
Comment (Dirli-V) [inline src/frontend/site.js]: [ISSUE] Please keep the spaces

File: src/frontend/site.js [hunk 5]
```diff
@@ -328,16 +341,65 @@ function postBio(req, res) {
...
+        fetchUsingDeploymentBase(req, () => req.LIKE_SERVICE_API.get(`/like-service/like-count/` + postId)
+        ).then((response => {
+            let countData = response.data;
+            resolve(countData);
+        }))
+    })
+
+}
+
+async function insertLikeCountIntoPostArray(req, data) {
```
Comment (Dirli-V) [inline src/frontend/site.js:377]: [DISCUSS] I think you could remove the outer promise:
```js
async function insertLikeCountIntoPostArray(req, data) {
    let finalData = [];
    await Promise.all(
        data.map(async (post) => {
            let postId = post.postId;
            let likeData = await getLikeCount(req, postId);
            let likeCount = likeData.likeCount;
            let userLiked = likeData.userLiked;
            post = { ...post, likeCount: likeCount, userLiked: userLiked };
            finalData.push(post);
        })
    );
    return finalData;
}
```

File: src/frontend/site.js [hunk 6]
```diff
@@ -328,16 +341,65 @@ function postBio(req, res) {
...
+}
+
+
+/**-
  * Creates a new Promise and performs base request before the one specified, this is useful to match the synchronous nature of
  * nunjucks, allowing microservice failures to not affect base requests. e.g deployment service requests will always succeed
  * regardless if login fails.
  *
- * @param req express req context
+ * @param req express req contextmy-timeline
```
Comment (Dirli-V) [inline src/frontend/site.js]: [ISSUE] Please revert the change :)

File: src/frontend/site.js [hunk 7]
```diff
@@ -328,16 +341,65 @@ function postBio(req, res) {
...
+            fetchUsingDeploymentBase(req, () => req.LIKE_SERVICE_API.post(`/like-service/like-post`, {postId: postId}
+            )).then((response => {
+                res.redirect(extendURL('/post/' + postId))
+            }))
+        }
+    }))
+}
+
+
+async function getLikeCount(req, postId) {
```
Comment (Dirli-V) [inline src/frontend/site.js:371]: [DISCUSS] I think you could remove the outer promise:
```js
async function getLikeCount(req, postId) {
    let response = await fetchUsingDeploymentBase(req, () => req.LIKE_SERVICE_API.get(`/like-service/like-count/` + postId))
    return response.data
}
```

File: src/frontend/site.js [hunk 8]
```diff
@@ -328,16 +341,65 @@ function postBio(req, res) {
...
             res.redirect(extendURL(`/user/${getJwtUser(req.cookies)}`));
         }).catch(error => {
-            res.status(statusCodeForError(error)).render('error.njk', handleError(error));
-        });
+        res.status(statusCodeForError(error)).render('error.njk', handleError(error));
+    });
 }
 
-/**
+function postLike(req, res) {
```
Comment (Dirli-V) [inline src/frontend/site.js]: [NEED_INFO] could we also make this function async? Then we can greatly decrease the nesting level here and improve readability.

File: src/frontend/site.js [hunk 9]
```diff
@@ -48,46 +48,54 @@ router.post('/bio/:username', postBio);
...
-function showPersonalTimeline(req, res) {
+async function showPersonalTimeline(req, res) {
     fetchUsingDeploymentBase(req, () =>
         Promise.all([
             req.MICROBLOG_API.get('/mytimeline'),
             getMembershipOfLoggedInUser(req)
         ]))
-        .then(([myTimeline, membership]) => {
+        .then(async ([myTimeline, membership]) => {
+            console.log("Start Data: " + myTimeline.data);
```
Comment (Dirli-V) [inline src/frontend/site.js]: [DISCUSS] I think we should remove these console.logs

File: src/frontend/site.js [hunk 10]
```diff
@@ -48,46 +48,54 @@ router.post('/bio/:username', postBio);
...
 
-function showPersonalTimeline(req, res) {
+async function showPersonalTimeline(req, res) {
     fetchUsingDeploymentBase(req, () =>
         Promise.all([
             req.MICROBLOG_API.get('/mytimeline'),
             getMembershipOfLoggedInUser(req)
         ]))
-        .then(([myTimeline, membership]) => {
+        .then(async ([myTimeline, membership]) => {
```
Comment (Dirli-V) [inline src/frontend/site.js]: [ISSUE] Please use await instead of then and reduce the nesting level here.

File: src/frontend/static/thumb_up.png:Zone.Identifier
Comment (Dirli-V) [inline src/frontend/static/thumb_up.png:Zone.Identifier:1]: [NEED_INFO] What is this file? :sweat_smile:

File: src/frontend/views/post.njk [hunk 11]
```diff
@@ -44,4 +44,23 @@ limitations under the License.
             </div>
         </div>
     </div>
+    <div class="card-footer">
+        <div style="float:right">
+                    <div style="display: inline-block;">
```
Comment (Dirli-V) [inline src/frontend/views/post.njk]: [ISSUE] Please reduce the indentation / reformat

File: src/frontend/views/post.njk [hunk 12]
```diff
@@ -44,4 +44,23 @@ limitations under the License.
...
+                        <p style="text-align:right; height:30px; font-size:25px;">{{post.likeCount}}</p>
+                    </div>
+                    {% if (post.userLiked) %}
+                        <form action="{{extendURL('/like/' + post.postId)}}" method="post" style="display: inline-block; height:46px">
+                            <input type="image" src="{{ extendURL('/img/thumb_up_liked.png') }}" width="30" height="30" style="vertical-align: bottom;"/>
+                      </form>
+                     {% endif %}
+                    {% if (true !== post.userLiked) %}
```
Comment (Dirli-V) [inline src/frontend/views/post.njk]: [NEED_INFO] Is there an `{% else %}` as well?

File: src/like-service/app/Http/Controllers/LikeController.php [hunk 13]
```diff
@@ -17,7 +17,9 @@ class LikeController extends BaseController
     function doLike($request)
     {
         $user_token = $request->cookie('jwt');
-        $postId = $request->header('postId');
+        $postId = $request->input('postId');
+
+        Log::notice($request->all());
```
Comment (Dirli-V) [inline src/like-service/app/Http/Controllers/LikeController.php]: [DISCUSS] I think we can remove this again

File: src/like-service/app/Http/Controllers/LikeController.php [hunk 14]
```diff
@@ -38,8 +40,11 @@ function doLike($request)
     function getLikeCountAndState($request)
     {
         $user_token = $request->cookie('jwt');
-        $postId = $request->header('postId');
+        $uri = $request->path();
+        $uriParts = explode('/', $uri);
+        $postId = $uriParts[2];
 
+        DB::enableQueryLog();
```
Comment (Dirli-V) [inline src/like-service/app/Http/Controllers/LikeController.php]: [DISCUSS] Also here, I think we can remove the logging. Same for the other logging logic here? Or does the route not provide the value for {postId}?

File: src/like-service/app/Http/Controllers/LikeController.php [hunk 15]
```diff
@@ -38,8 +40,11 @@ function doLike($request)
     function getLikeCountAndState($request)
     {
         $user_token = $request->cookie('jwt');
-        $postId = $request->header('postId');
+        $uri = $request->path();
```
Comment (Dirli-V) [inline src/like-service/app/Http/Controllers/LikeController.php:40]: [NEED_INFO] Can not just use GET parameters here?

File: src/microblog-service/src/main/java/org/dynatrace/microblog/MicroblogController.java [hunk 16]
```diff
@@ -126,7 +126,9 @@ public List<Post> getUserPosts(@PathVariable("user") String user,
 
         if (!userAuthServiceClient.checkTokenValidity(jwt)) throw new InvalidJwtException();
 
-        return redisClient.getUserPosts(jwt, user, Integer.parseInt(limit));
+        List<Post> postList = redisClient.getUserPosts(jwt, user, Integer.parseInt(limit));
+        logger.info(postList.get(1).getPostId());
```
Comment (Dirli-V) [inline src/microblog-service/src/main/java/org/dynatrace/microblog/MicroblogController.java]: [ISSUE] Please revert this again

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java [hunk 17]
```diff
@@ -25,16 +25,19 @@ public class Post {
...
     private final String imageUrl;
+    private final String postId;
 
     public Post(
 			@JsonProperty("username") String username,
 			@JsonProperty("body") String body,
 			@JsonProperty("imageUrl") String imageUrl,
-			@JsonProperty("timestamp") Date timestamp) {
+			@JsonProperty("timestamp") Date timestamp,
+            @JsonProperty("postId") String postId){
```
Comment (Dirli-V) [inline src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java]: [ISSUE] Please reformat

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java [hunk 18]
```diff
@@ -52,4 +55,6 @@ public Date getTimestamp() {
     public String getImageUrl() {
         return imageUrl;
     }
+
+    public String getPostId(){ return postId; };
```
Comment (Dirli-V) [inline src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java]: [ISSUE] Please reformat

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/SerializedPost.java [hunk 19]
```diff
@@ -30,8 +30,9 @@ public SerializedPost(
 			@JsonProperty("body") String body,
 			@JsonProperty("imageUrl") String imageUrl,
 			@JsonProperty("timestamp") Date timestamp,
+            @JsonProperty("postId") String postId,
```
Comment (Dirli-V) [inline src/microblog-service/src/main/java/org/dynatrace/microblog/dto/SerializedPost.java]: [ISSUE] Please reformat

**Review by Dirli-V — CHANGES_REQUESTED** [review]

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/SerializedPost.java (hunk 19 above)
Comment (KugliLuca) [reply src/microblog-service/src/main/java/org/dynatrace/microblog/dto/SerializedPost.java]: Reformatted

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java (hunk 18 above)
Comment (KugliLuca) [reply src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java]: Reformatted

File: src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java (hunk 17 above)
Comment (KugliLuca) [reply src/microblog-service/src/main/java/org/dynatrace/microblog/dto/Post.java]: Reformatted

File: src/microblog-service/src/main/java/org/dynatrace/microblog/MicroblogController.java (hunk 16 above)
Comment (KugliLuca) [reply src/microblog-service/src/main/java/org/dynatrace/microblog/MicroblogController.java]: Reverted

File: src/frontend/site.js (hunk 6 above)
Comment (KugliLuca) [reply src/frontend/site.js]: Reverted

File: src/frontend/site.js (hunk 9 above)
Comment (KugliLuca) [reply src/frontend/site.js]: Removed

File: src/like-service/app/Http/Controllers/LikeController.php (hunk 13 above)
Comment (KugliLuca) [reply src/like-service/app/Http/Controllers/LikeController.php]: Removed

File: src/like-service/app/Http/Controllers/LikeController.php (hunk 14 above)
Comment (KugliLuca) [reply src/like-service/app/Http/Controllers/LikeController.php]: Logging logic removed

File: src/frontend/app.js (hunk 3 above)
Comment (KugliLuca) [reply src/frontend/app.js:148]: Removed

File: src/frontend/site.js (hunk 4 above)
Comment (KugliLuca) [reply src/frontend/site.js]: Spaces re-added

File: src/frontend/site.js (hunk 8 above)
Comment (KugliLuca) [reply src/frontend/site.js]: Function made Async and nesting level reduced.

File: src/frontend/views/post.njk (hunk 11 above)
Comment (KugliLuca) [reply src/frontend/views/post.njk]: Reformatted

File: src/frontend/views/post.njk (hunk 12 above)
Comment (KugliLuca) [reply src/frontend/views/post.njk]: Else instead of if now used

File: src/frontend/.env (hunk 1 above)
Comment (KugliLuca) [reply src/frontend/.env]: Line removed

File: src/frontend/app.js (hunk 2 above)
Comment (KugliLuca) [reply src/frontend/app.js]: Todo comment added

File: src/frontend/static/thumb_up.png:Zone.Identifier
Comment (KugliLuca) [reply src/frontend/static/thumb_up.png:Zone.Identifier:1]: File removed

File: src/frontend/site.js (hunk 7 above)
Comment (KugliLuca) [reply src/frontend/site.js:371]: Outer promise removed

File: src/frontend/site.js (hunk 10 above)
Comment (KugliLuca) [reply src/frontend/site.js]: Await used and nesting level reduced

File: src/like-service/app/Http/Controllers/LikeController.php (hunk 15 above)
Comment (KugliLuca) [reply src/like-service/app/Http/Controllers/LikeController.php:40]: We are. It's easiest to get to them through the URI

**Review by Dirli-V — APPROVED** [review]
